src/data_source/storage/sql.rs

```diff
@@ -1170,6 +1192,401 @@ where
     }
 }
 
+#[async_trait]
+impl UpdateStateData for SqlStorage {
+    type Error = QueryError;
+    async fn insert_merkle_nodes<
+        Proof: MembershipProof<E, I, T> + Send + Sync + Debug + 'static,
+        E: Element + Send + Sync + Serialize,
+        I: Index + Send + Sync + Serialize,
+        T: NodeValue + Send + Sync,
+    >(
+        &mut self,
+        name: &'static str,
+        proof: Proof,
+        traversal_path: Vec<usize>,
+        block_number: u64,
+    ) -> QueryResult<()> {
+        let pos = proof.index().clone();
+        // First node in the merkle path is always the leaf node for which membership proof is provided
+        let Some(MerkleNode::Leaf { value, elem, .. }) = proof.merkle_path().first() else {
+            return Err(QueryError::Error {
+                message: "membership proof has a missing leaf".to_owned(),
+            });
+        };
+
+        let mut leaf_commit = Vec::new();
+        // Serialize the leaf node hash value into a vector
+        value
+            .serialize_compressed(&mut leaf_commit)
+            .map_err(ParseError::Serialize)?;
+
+        // Define the LTREE path to be inserted for the leaf node.
+        // Note:
+        // The traversal path is from the leaf to the root.
+        // The LTREE path is created by appending 'R' for the Root node to this path and then reversing it.
```

Comment:
Why do we need to represent the root explicitly? It's a common ancestor of every path so could just be represented by the empty ltree.